src/Lean/Widget/InteractiveDiagnostic.lean

```diff
@@ -55,31 +60,45 @@ private def mkPPContext (nCtx : NamingContext) (ctx : MessageDataContext) : PPCo
   currNamespace := nCtx.currNamespace, openDecls := nCtx.openDecls
 }
 
+/-! The `msgToInteractive` algorithm turns a `MessageData` into `TaggedText MsgEmbed` in two stages.
+
+First, in `msgToInteractiveAux` we produce a `Format` object whose `.tag` nodes refer to `EmbedFmt`
+objects stored in an auxiliary array. Only the most shallow `.tag` in every branch through the
+`Format` corresponds to an `EmbedFmt`. The kind of this tag determines how the nested `Format`
+object (possibly including further `.tag`s), is processed. For example, if the output is
+`.tag (.expr ctx infos) fmt` then tags in the nested `fmt` object refer to elements of `infos`.
+
+In the second stage, we recursively transform such a `Format` into `TaggedText MsgEmbed` according
+to the rule above by first pretty-printing it and then grabbing data referenced by the tags from
+all the nested arrays (such as the `infos` array in the example above).
+
+We cannot easily do the translation in a single `MessageData → TaggedText MsgEmbed` step because
+that would effectively require reimplementing the (stateful, to keep track of indentation)
```

Comment:
@Vtec234 The pretty printer does not track indentation. Only `Format.pretty` does, but I guess that's not what you mean here.